examples/trusted_information_retrieval/module_1/rust/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+    };
+    client.get_database_entry(request).and_then(|response| {
+        response
+            .entry
+            .ok_or_else(|| grpc::build_status(grpc::Code::Internal, "Malformed response"))
+    })
+}
+
+/// Oak Node that connects to an external database.
+#[derive(Default)]
+pub struct DatabaseProxyNode;
+
+impl DatabaseProxyNode {
+    /// Create a gRPC client pseudo-Node.
+    fn get_grpc_client(database_url: &str) -> anyhow::Result<DatabaseClient> {
+        oak::grpc::client::Client::new_with_label(
+            &oak::node_config::grpc_client(database_url),
+            &Label::public_untrusted(),
+        )
+        .map(DatabaseClient)
+        .context("Couldn't create a gRPC client")
+    }
+
+    /// Load a database subset defined by `offset` and the number of requested elements
+    /// (`page_size`).
+    fn list_database_entries(
+        offset: u32,
+        page_size: u32,
+        database_url: &str,
+    ) -> anyhow::Result<Vec<DatabaseEntry>> {
+        let request = ListDatabaseEntriesRequest {
+            offset: offset as i32,
+            page_size: page_size as i32,
+        };
+        Self::get_grpc_client(database_url)?
+            .list_database_entries(request)
+            .map(|response| response.entries)
+            .map_err(|error| anyhow!("gRPC send error: {:?}", error))
+    }
+}
+
+/// A gRPC service implementation for the database proxy Node.
+impl DatabaseProxy for DatabaseProxyNode {
+    fn get_database_entry(
+        &mut self,
+        request: GetDatabaseEntryRequest,
+    ) -> grpc::Result<GetDatabaseEntryResponse> {
+        // Request entries from the database until the last entry page is reached.
+        let mut requested_entry = None;
+        let mut current_offset = 0;
+        loop {
+            let entries = Self::list_database_entries(
+                current_offset,
+                DATABASE_PAGE_SIZE,
+                &request.database_url,
+            )
+            .map_err(|error| {
+                grpc::build_status(
+                    grpc::Code::NotFound,
+                    &format!("Database error: {:?}", error)[..],
+                )
+            })?;
+
+            // When the requested entry is found, we continue requesting new database entries, so
+            // the database doesn't learn, which entry was requested.
+            if let Some(entry) = entries.iter().find(|e| e.key == request.key) {
+                requested_entry = Some(entry.clone());
+            }
+
+            // If the response contains less entries than expected, then it's the last database
+            // page.
+            if entries.len() < DATABASE_PAGE_SIZE as usize {
+                break;
+            }
+            current_offset += DATABASE_PAGE_SIZE;
+        }
+
+        requested_entry
+            .map(|entry| GetDatabaseEntryResponse { entry: Some(entry) })
+            .ok_or_else(|| grpc::build_status(grpc::Code::NotFound, ""))
+    }
```

Comment:
IIUC, this will return an error to the caller, but not log anything in the server, which may make debugging hard / impossible. Or does it get logged somewhere else?